### DISCUSSION TITLE
Debuging W806 in Arduino IDE 2.0.3

### DISCUSSION THREAD

**Discussion by AnatolSher** [board707/w80x_arduino#9] [Ideas]
На всякий случай дернул у T-Head сервер отладки для Win и Linux. Самодельный CKLink-Lite видит, порты открывает... не ругается
Инструкцию по применению придется переводить с китайского.

Just in case, I pulled the debug server for Win and Linux from T-Head. Homemade CKLink-Lite sees, opens ports ... does not swear
Instructions for use will have to be translated from Chinese.

![image](https://user-images.githubusercontent.com/121094544/216779205-d70ed974-5dfa-42fa-953a-eed7ec629ce5.png)
[T-Head-DebugServer-windows-V5.12.11-20211229-1605.zip](https://github.com/board707/w80x_arduino/files/10609104/T-Head-DebugServer-windows-V5.12.11-20211229-1605.zip)
[T-Head-DebugServer-linux-x86_64-V5.12.11-20211229.sh.tar.gz](https://github.com/board707/w80x_arduino/files/10609106/T-Head-DebugServer-linux-x86_64-V5.12.11-20211229.sh.tar.gz)

UPD:
В рукопашную отладка работает! Надо разбираться как в IDE 2/0/3 ее вкрутить. Я еще не все команды GDB изучил. 
Видно какой то косяк... Брекпойнт не установился.

In hand mode, debugging via C-SKY GDB works. We need to figure out how to add it to IDE 2/0/3
![image](https://user-images.githubusercontent.com/121094544/216783388-8269fd53-d1a6-4325-bcc4-d53cd1b75986.png)

**Comment by AnatolSher** [comment]
Проверил как ведет себя сервер отладки и отладчик в ОС Linux. Программы запускаются


![Снимок экрана от 2023-02-06 19-21-22](https://user-images.githubusercontent.com/121094544/217026791-9acfd30e-0205-46a0-84db-1db6f4688c29.png)
![cklink_mapping](https://user-images.githubusercontent.com/121094544/217031568-35a8a88b-b163-4427-8be7-e750d71b40d5.png)

Я сейчас работаю с Linux в виртуальной среде VBox. Последний релиз x64 Fedora
СКLink-Lite успешно замапился.
Нужно подгрузить ncurses-compat-lib (sudo yum install ncurses-compat-libs)
csky-elfabiv2-gdb отладчик лежит в папке .Arduino15/packages/w80x_duino/tools/csky/2021.04.23/bin

I checked how the debug server and debugger behave in Linux OS. Programs run
I'm currently working with Linux in a VBox virtual environment. Latest x64 Fedora release
CKLink-Lite successfully mapped.
You need to download ncurses-compat-lib (sudo yum install ncurses-compat-libs)
csky-elfabiv2-gdb debugger is located in the folder .Arduino15/packages/w80x_duino/tools/csky/2021.04.23/bin

**Comment by AnatolSher** [comment]
**Рекомендация по сборке CKLink-Lite**
В схеме у henrycoder обозначен контроллер STM32F103C8T6. Однако, с этим контроллером я получил не стабильную работу.
По моему мнению нужен контроллер не менее 128kb флеш памяти. Например STM32F103CBT6. Такой "голубой таблетки" у меня не было. В дело пошел неизвестный китайский клон с 256kb на борту.

**CKLink-Lite build recommendation**
In the henrycoder's schematic diagramhas an STM32F103C8T6 controller present. However, with this controller, I did not get stable work.
In my opinion, you need a controller with at least 128kb of flash memory. For example STM32F103CBT6. I didn't have that original "blue pill". An unknown Chinese clone with 256kb on board went into action.

**Comment by board707** [comment]
Спасибо за информацию.
Честно говоря, пока внимательно в отладку не вникал.
Не могли бы вы расписать по шагам, как установить и запустить отладчик с нашей платой?
Или если где-то есть такое описание - дайте, пожалуйста, ссылку.

**Reply by AnatolSher** [reply]
На T-Head Debug Server делаю перевод с китайского. Может за пару недель управлюсь 
Я еще здесь немного касаюсь  темы CDK и отладки с самодельным CKLink-Lite
https://radiokot.ru/forum/viewtopic.php?f=62&t=185076

**Reply by AnatolSher** [reply]
Оооопс! Повезло :) Нашел туториал на английском!
[T-Head Debugger Server User Guide (EN-US).pdf](https://github.com/board707/w80x_arduino/files/10676212/T-Head.Debugger.Server.User.Guide.EN-US.pdf)
Что то сегодня прет не по децки :)
[C-SKY+Tools+V3+User+Guide-GDB.pdf](https://github.com/board707/w80x_arduino/files/10677352/C-SKY%2BTools%2BV3%2BUser%2BGuide-GDB.pdf)

**Reply by AnatolSher** [reply]
> Не могли бы вы расписать по шагам, как установить и запустить отладчик с нашей платой? Или если где-то есть такое описание - дайте, пожалуйста, ссылку.

Если Вы имеете в виду саму железяку CKLink-Lite, то вот здесь расписано все
https://hackaday.io/page/12671-how-to-debug-w806-and-w801-mcus

**Comment by AnatolSher** [comment]
Вот нашел кино как пользоваться олдскульной отладкой GDB на примере STM
https://www.youtube.com/watch?v=D-MbC9N5U7o&t=558s
Но так пока у меня не получилось, не смотря на то, что основные шаги были выполнены успешно. 
Китайцы отсылают к dbg.org за туториалом. Читал... долго думал :)  Но что то узкоглазые не договаривают

**Comment by AnatolSher** [comment]
Наткнулся на странную вещь! Вот что выдал мне gdb отладчик
![image](https://user-images.githubusercontent.com/121094544/217645896-8f5a3e38-239a-4e4a-aea6-3c8840d59238.png)
Откуда в elf файле ссылка на несуществующий маршрут на моем компьютере?
Возможно где то лежит сценарий в пакете 80x-duino. Но я его найти не могу. Отладчик затыкается на этом месте
Там вроде написано что нет такого файла. Но у меня и диска D: нету

**Comment by board707** [comment]
По-моему у товарища nulllab лежали прекомпилированные версии всех .S файлов в его ардуино-пакете, видимо собранные на его компе.
 Может их удалить и собрать заново?

**Comment by AnatolSher** [comment]
Я попробую грохнуть все с расширением *.o и перекомпилировать скетч

**Reply by AnatolSher** [reply]
Грохнул! Вот что теперь получается
![image](https://user-images.githubusercontent.com/121094544/217876576-5e519b29-bc10-4280-b9ba-5f57e1435b2b.png)
Как заставить ардуинку всегда заново пересобирать ядро?
А то я потерялся уже... :)

**Reply by board707** [reply]
>  Как заставить ардуинку всегда заново пересобирать ядро?

~~Этого не знаю.~~

Дополнение - нашел, кроме стирания `.o` файлов, нужно еще переименовать соответвующие исходники. Они там лежат с расширениями  `.S_` - специально, чтобы не пересобирались каждый раз. Их надо поменять на  `.S`
Только бестолку - сейчас попробовал, не собираются, каких-то других файлов не хватает.

**Reply by AnatolSher** [reply]
Здается мне что вот тут собака зарыта
# These can be overridden in platform.local.txt
compiler.c.extra_flags=
compiler.c.elf.extra_flags=-Wl,--no-whole-archive -nostartfiles -mhard-float
#compiler.S.extra_flags=-I{runtime.tools.csky.path}/csky-elfabiv2/include/csi/csi_core/csi_cdk/ -ID:/C-Sky/CDK/CSKY/csi/csi_core/include/ -ID:/C-Sky/CDK/CSKY/csi/csi_driver/include/
compiler.S.extra_flags=
compiler.cpp.extra_flags=
compiler.ar.extra_flags=
compiler.objcopy.eep.extra_flags=
compiler.elf2bin.extra_flags=
compiler.includes="-I{runtime.tools.csky.path}/csky-elfabiv2/include" "-I{build.core.path}/include" "-I{build.core.path}/include/arch/{build.arch}" "-I{build.core.path}/include/arch/{build.arch}/csi_dsp" "-I{build.core.path}/include/arch/{build.arch}/csi_core" "-I{build.core.path}/include/driver" 
compiler.libs={build.core.path}/lib/libdsp.a {build.core.path}/lib/arch/{build.arch}/bsp/startup.o {build.core.path}/lib/arch/{build.arch}/bsp/__rt_entry.o {build.core.path}/lib/arch/{build.arch}/bsp/vectors.o

В нижней строке берутся объектные файлы
А в закомментированной строке идет отсылка на CDK... не SDK! Попробую там пошаманить

**Comment by board707** [comment]
Я думаю можно стереть эти `.o` файлы в SDK и попробовать собрать какой-нить из встроенных примеров. И посмотреть, как и с какими опциями SDK их соберет. Потом перенести в ардуино.
В SDK они заведомо собираются, при скачивании с гитхаба там обьектных файлов нет.

Добавка: - а можно просто запустить сборку в СДК и потом положить получившиеся файлы в ардуино пакет, как это сделал товарищ nulllab

**Reply by AnatolSher** [reply]
Да нееее... Это все разруливается в Platform.txt
После рихтовки параметра 
compiler.S.extra_flags= -I{build.core.path}/include -Ibuild.core.path}/include/arch/{build.arch} -I{build.core.path}/include/arch/{build.arch}/csi_dsp -I{build.core.path}/include/arch/{build.arch}/csi_core -I{build.core.path}/include/driver -I{build.core.path}/include/arch/{build.arch} 
Компилятор *.S файлы видит, но почему то  не создает на них объектники с расширением *.o и *.o.d
Эта фаза зависит наверное от другого параметра
## Compile S files
recipe.S.o.pattern="{compiler.path}{compiler.c.cmd}" -c "{source_file}" -mcpu={build.mcpu} {compiler.S.flags} -MMD -MP -MT{object_file} -MF{object_file}.d -o "{object_file}" {compiler.S.extra_flags} {build.extra_flags} {compiler.includes} {includes}

Тут я пока ничего не понимаю

**Comment by board707** [comment]
Чтобы компилятор увидел .S файлы, никаких параметров в настройках "рихтовать" не надо. У меня они видятся сразу.
А для создания .o.d файлов в ардуино просто нет рецепта, вот они и не создаются.
Поэтому я предлагаю посмотреть, как они создаются в СДК - а потом это перенести в наш пакет. На мой взгляд, это самый простой путь.
А подбирать параметры , тем более не очень понимая, что они значат - тупиковый путь.

**Reply by AnatolSher** [reply]
> Чтобы компилятор увидел .S файлы, никаких параметров в настройках "рихтовать" не надо. У меня они видятся сразу. 


Это я уже заметил :)
А выдернуть что либо из CDK это вообще mission imbosible
Я этот CDK не полных пару месяцев в руках верчу. Только разобрался с самыми примитивными задачами
Мне кажется что в рецепте для S файлов нужно просто заставить компилятор писать объектник в ту же папку где он взял S. Т.е затирал предыдущие сеансы. Бикоз следом стоит опция подхватывать эти объектники как библиотечные файлы.

**Reply by AnatolSher** [reply]
> А подбирать параметры , тем более не очень понимая, что они значат - тупиковый путь.

Как когда то говорили на нашей кафедре - метод научного тыка :) Иногда срабатывает

**Reply by board707** [reply]
> выдернуть что либо из CDK это вообще mission imbosible

Мне кажется Вы меня не поняли. Ничего выдергивать не надо. У вас же запускаются примеры из папки demo в SDK ? - должны! это работает сразу, из коробки. 
Сотрите обьектные файлы и запустите сбору любого примера - компилятор вам сам покажет, с какими опциями собираются эти файлы.

**Reply by AnatolSher** [reply]
> А для создания .o.d файлов в ардуино просто нет рецепта, вот они и не создаются. 

Рецепт есть, только я не понимаю как его поперчить/посолить. Вот здесь он
## Compile S files
recipe.S.o.pattern="{compiler.path}{compiler.c.cmd}" -c "{source_file}" -mcpu={build.mcpu} {compiler.S.flags} -MMD -MP -MT"{object_file}" -MF"{object_file}.d" -o "{object_file}" {compiler.S.extra_flags} {build.extra_flags} {compiler.includes} {includes}

**Reply by AnatolSher** [reply]
> Мне кажется Вы меня не поняли. Ничего выдергивать не надо. 

Теперь понял :)

**Reply by AnatolSher** [reply]
В общем... Мне удалось заставить компилировать без ошибок. Но толку от этого мало... к сожалению
Если убрать китайские объектники и собрать прошивку, то она не стартует на плате. Похоже этот nullaborg собрал ядро отдельно и исходников не оставил
И еще одно наблюдение. В исходном SDK вообще нет __rt_entry.S И размеры остальных вроде отличаются. На содержимое не проверял. 

Вот так происходит сборка в CDK
![image](https://user-images.githubusercontent.com/121094544/217940645-94a5f5be-d07f-4c34-a35d-22f42b7fdd34.png)
Я так и сделал в Platform.txt
Но явно __rt_entry.S я не обнаружил. Ни в CDK ни в SDK

Но еще поковыряюсь

**Reply by board707** [reply]
> Вот так происходит сборка в CDK ![image](https://user-images.githubusercontent.com/121094544/217940645-94a5f5be-d07f-4c34-a35d-22f42b7fdd34.png) 
Я так и сделал в Platform.txt 

Выложите куда-нибудь Platform.txt , который у вас получился. Только не скриншотом! Вообще никогда не выкладывайте команды консоли скриншотом(как выше), пожалуйста.

**Reply by AnatolSher** [reply]
Это был вывод из CDK. Я пока не знаю куда он логи компиляции складывает

**Comment by board707** [comment]
Не, надо копать дальше. Раз в SDK все собирается и потом работает - значит там все есть для успешной сборки. Вывод - и в ардуино должно собираться.
И, кстати. если вы собираете прошивку для дебага - вам и обьекты из файлов .S надо собирать с ключиками отладки, а не так как прописано в SDK

**Reply by AnatolSher** [reply]
-g3 -ggdb я ставил

**Comment by board707** [comment]
Сейчас попробовал - у меня все собирается.

1.Распаковал SDK, взятый по [Вашей ссылке](https://github.com/IOsetting/wm-sdk-w806)
2.Настроил путь к компилятору через `make menuconfig`
3. `make lib`
4. `make all`

После этого в каталоге `bin/build/W806/obj/platform/arch/xt804/bsp` будут лежать интересующие нас `.o` файлы. Их можно взять и скопировать в пакет для ардуино. С этими новыми обьектными файлами у меня все собирается и даже работает. 
Если нужен файл `__rt_entry.o` - то его можно сгенерить в том же SDKиз того исходника, что нам оставил товарищ nulllab. Для этого просто подложите файл `__rt_entry.S` в тот каталог SDK где лежали другие .S файлы и повторить пункты 2 и 3 выше.


PS Выложите куда-нибудь  свой Platform.txt, куда вы добавляли рецепт для сборки .о файлов. Только не скриншотом.

**Reply by AnatolSher** [reply]
Чуть позже... вечером или ночью

На всякий случай
[C-SKY+Tools+V3+User+Guide-Compiler.pdf](https://github.com/board707/w80x_arduino/files/10708176/C-SKY%2BTools%2BV3%2BUser%2BGuide-Compiler.pdf)
[C-SKY+Tools+V3+User+Guide-Link.pdf](https://github.com/board707/w80x_arduino/files/10708180/C-SKY%2BTools%2BV3%2BUser%2BGuide-Link.pdf)

Ну и стандарт

[C-SKY+ABIV2+Standards+Manual.pdf](https://github.com/board707/w80x_arduino/files/10708196/C-SKY%2BABIV2%2BStandards%2BManual.pdf)
[C-SKY+ABIV1+Standards+Manual.pdf](https://github.com/board707/w80x_arduino/files/10708199/C-SKY%2BABIV1%2BStandards%2BManual.pdf)

**Comment by AnatolSher** [comment]
Ну вот... platform.txt
Компилирует без ошибок... но лыжи не едут :)
[platform.txt](https://github.com/board707/w80x_arduino/files/10709378/platform.txt)
В результате компиляции все объектники попадают в /tmp и там собираются воедино
![image](https://user-images.githubusercontent.com/121094544/218144567-fd66d83c-ede5-45dc-a8ae-220e156f7d1e.png)

А это протокол компиляции из CDK IDE


[BuildLog.txt](https://github.com/board707/w80x_arduino/files/10709474/BuildLog.txt)

Возможно я что то упустил... Но уже глаз замылился. И меня уже колбасит не по детцки :) из-за этой проблемы
На сколько я понял, *.S файлы это ассемблерные вставки и они компилируются с такими же опциями как и обычные *.c

**Comment by board707** [comment]
Ок, я тоже это все проделал. Я даже выставил абсолютно одни и те же опции компиляции при сборке под ардуино и непосредственно в SDK. Все собирается без ошибок, но с обьекниками из SDK прошивка работает, а из ардуино нет.
 
Предлагаю вернутся к исходному вопросу. Зачем нам нужно пересобирать эти файлы? Если речь о том, чтобы обеспечить работу отладки, то не можем мы скомпилировать правильные файлы в SDK, а потом просто подложить их в Ардуино? Это не решит проблему?

**Reply by AnatolSher** [reply]
Нам нужно что бы эти объектники были гарантировано с ключом -g3 и были доступны их реальные исходники. Иначе gdb работать как нам хочется не будет.
Я еще попробую воссоздать рабочее окружение nulllaborg-а. Даже еще винт прицепил на комп. Назначил букву D:
Этот товарищ использовал CDK, вот я и подсуну маршрут который прописан в этих объектниках. Посмотрим что получится. 
Возможно, то что Вы предлагаете - сработает. Но меня мучает вопрос... какого хрена там делает __rt_entry.S и __rt_entry.o? Если в стандартном SDK после компиляции этого файла не видно... зачем nulllaborg его туда подсунул?
Среда CDK IDE этот файлы генерирует и эти файлы присутствуют в папке /Obj и они участвуют в окончательной сборке прошивки, хотя в логах их следа нет. И тоже все работает. Как то так. Чудеса...

**Comment by board707** [comment]
Не похоже что в этом дело.
В SDK они собираются без этого маршрута и работают...

Ладно, удачи.
Я пока буольше этим заниматься не буду, отладка это явно не первой важности цель в этом проекте :)

**Reply by AnatolSher** [reply]
Без отладки будет очень сложно воссоздать стандартный ардуиновский набор Wiring. А потом еще согласовать вызовы внешних библиотек типа Адафруит и т.д Тут printf и морганием светодиодов не обойдешься.
Я еще одну попытку сделаю... Не получится, возьму перерыв :)
Какая задача первоочередная у Вас? Я думаю что надо сборку 0.0.2с сделать с новым Hal

**Reply by AnatolSher** [reply]
Вот подумал... может версия тулчейна имеет значение? nulllaborg это делал на старье

**Reply by AnatolSher** [reply]
> Не похоже что в этом дело. В SDK они собираются без этого маршрута и работают...

Маршрут нужен для отладчика. По этому маршруту находятся исходники

**Comment by board707** [comment]
Версия тулчейна не может влиять, потому что файлы, созданные этим новым тулчейном в SDK - работают в ардуино! 

Похоже я нашел в чем дело, но пока не знаю, как поправить. Среда ардуино создает обьектные файлы с именем xxxx.S.o и это имя сохраняется в списке обьектных файлов. Однако при сборке библиотеки, похоже, использовано имя xxxx.o и потому часть символов из обьектных файлов не разрешается.
Мне кажется, проблема уйдет, если заставить ардуино ИДЕ собирать эти файлы с именем xxxx.o. Но в рецепте имя выходного файла генерится автоматически и я пока не нашел, как это изменить.

**Reply by AnatolSher** [reply]
-MMD -MP -MF {object_file}.d -MT {object_file} -o "{object_file}"
Эти опции управляют компилятором и говорят с каким расширением будут выходные файлы
Например
If -o is not specified, the default is to put an executable file in a.out, the object file for source.suffix in source.o, its assembler file in source.s, a precompiled header file in source.suffix.gch, and all preprocessed C source on standard output.
Т.е при компиляции на входе целый винигрет с разными расширениями одного и того же исходника. А вот что линковщик (если такое есть) собирает... это вопрос.

**Reply by board707** [reply]
> Эти опции управляют компилятором и говорят с каким расширением будут выходные файлы 

Это я в курсе. Но в рецептах сборки platform.txt это не работает.

В итоге у меня все собирается. Я был прав насчет того, что дело было в именах файлов - проблема в том, что библиотека gcc_csky.ld  содержит ссылку на startup.o. 

В итоге создал на каждый из трех .S файлов свой рецепт с генерацией обьектного файла `.o`, а не `.S.o` . Далее возникла проблема, как не дать ардуине откомпилировать файлы еще раз -  пришлось спрятать исходные `.S`  файлы в отдельный каталог снаружи дерева core. Но в итоге теперь все собирается и прошивка работает.

**Comment by board707** [comment]
Вот вам мой файл
[platform.txt](https://github.com/board707/w80x_arduino/files/10711989/platform.txt)
В том каталоге, где лежит platform.txt  - создайте директорию slib и скопируйте в нее три наших .S файла. В том каталоге, где они лежали изначально, переименуйте их .S_ - чтобы не подхватывались компилятором.
Для чистоты экперимента соответвующие обьектные файлы можете убрать, чтобы быть увренным, что они сгенерятся заново.
Удачи.

**Reply by AnatolSher** [reply]
Вот здорово! Спасибо

И еще один "наглый" вопрос :)
Мы можем заставить ардуино иде создавать символические ссылки во временную папку, где собирается прошивка и *.elf файл?
Для дальнейшего юзабилити там должны быть ссылки на два файла. T-Head debug server и cskygdb.
Папка T-Head возникает после инсталяции пакета (см.выше) в домашнем каталоге.

**Reply by board707** [reply]
> Мы можем заставить ардуино иде создавать символические ссылки во временную папку, где собирается прошивка и *.elf файл? 

Где эти ссылки должны располагаться? В пользовательском каталоге где скетч?
Определить временную папку, где ИДЕ собирает прошивку - не проблема. А вот для каталога скетча такой переменной сходу не нашел. Если найду, то под линуксом создать ссылку не проблема, под виндой надо пробовать.

**Reply by AnatolSher** [reply]
Нет. Они должны быть во временной папке где собран *.elf (/tmp)
Это наш предмет отладки

**Reply by board707** [reply]
> Нет. Они должны быть во временной папке где собран *.elf (/tmp) Это наш предмет отладки

У ссылки, так сказать, два конца. Один - где она расположена, а второй - куда она указывает.  Я понял, что один из концов - во временной папке где ардуино собирает проект. А второй?

**Reply by AnatolSher** [reply]
gdb лежит здесь packages\w80x_duino\tools\csky\2021.04.23\bin
T-Head debug server будет лежать в домашнем каталоге после ручной инсталляции под $User
Но, наверное, было бы правильно включить debug server в состав пакета w80x_duino
Единственно что нужно учесть что эта штука имеет зависимость от ncurses-compat-lib (sudo yum install ncurses-compat-libs)
Тогда вообще никаких проблем. 

Но в принципе этот пункт избыточен на текущем этапе. Можно ручками запустить debug server (он же прокси между отладчиком CKLink-Lite и gdb)
Но на будущее... когда будем внедрять отладку в Arduino IDE 2х это будет актуально. Т.е T-Head debug server в составе пакета

**Comment by board707** [comment]
С утра нашел более простое и изящное решение. Не надо никаких рецептов, не надо выкладывать файлы в отдельный каталог... Достаточно открыть файл  `cores/w806/ld/gcc_csky.ld` любым текстовым редактором (это текстовой файл), найти в нем две строки, где встречается имя файла `startup.o` и исправить его на `startup.S.o` .
После этого все работает

**Reply by AnatolSher** [reply]
Я вообще подумал что надо сломать схему сборки ардуиновскую с ее временным каталогом. Всю сборку перенаправить в папку где лежит скетч. Создав там каталог для сборки /build
Но... люди ко всему привыкают :)

**Reply by AnatolSher** [reply]
> С утра нашел более простое и изящное решение. 

В итоге результирующий platform.txt как выглядит?

**Comment by board707** [comment]
Не думаю, что это правильный путь. Все-таки мы (по крайней мере я) хотим создать пакет для Ардуино. Чем ближе мы будем придерживаться ардуино стандартов, тем проще будет интегрировать пакет в среду и тем меньше будет потом проблем с его поддержкой.

**Comment by board707** [comment]
Файлы
[platform.txt](https://github.com/board707/w80x_arduino/files/10713440/platform.txt)
[gcc_csky.ld.txt](https://github.com/board707/w80x_arduino/files/10713444/gcc_csky.ld.txt)
У второго удалите двойное расширение .тхт, это пришлось добавить, а то форма сообщения такого файла не понимает

**Reply by AnatolSher** [reply]
Ok. Спасибо

**Comment by board707** [comment]
Анатолий, по отладке это вы решайте сами, я в нее не вникал. Если надо будет линк на каталог сборки - пишите, сделаю.

Что по последним изменением - помогла Вам перекомпиляция .S файлов? Gdb начал запускаться?

**Reply by AnatolSher** [reply]
Да! Все прекрасно компилируется, прошивка стартует на плате. Дальше копаю gdb, все запускается. 
Возможно, еще с ключами компиляции придется повозиться. Мне кажется есть немного различий в том что делают CDK IDE и Arduino IDE
Исходники стартапа видит! Это уже очень хорошо.
![image](https://user-images.githubusercontent.com/121094544/218273838-620e685e-ff39-469f-a2d9-52ba2fc294df.png)

**Reply by AnatolSher** [reply]
Вот объяснение первых граблей... китайцы из WinnerMicro отвечают
http://ask.winnermicro.com/question/127.html
на это предупреждение
![image](https://user-images.githubusercontent.com/121094544/218275138-3f658041-5749-4743-a2cd-f47bb8760828.png)
Интересно а в коде ядра есть функция типа __BKPT()

**Comment by board707** [comment]
Это вы им там вопросы задаете?

**Reply by AnatolSher** [reply]
:) Мы не первые и не последние кто на эти грабли наступил. Все поиском
__BKPT() я нашел... но он не сработал
Вот тут малопонятное объяснение как рулить памятью камня.
http://ask.winnermicro.com/question/19.html
В теории мы можем указать что должно работать из ОЗУ. Тогда мягкие брекпойнты можно устанавливать из gdb. Но может я ошибаюсь.

Проблема в том что на WinnerMicro нет возможности зарегистрироваться если ты не житель поднебесной. 
Надо искать китайца и привлекать к проекту :)

**Reply by AnatolSher** [reply]
В общем... пришел к выводу что надо научиться рулить памятью камня. Если посмотреть что в *.elf написано, то видно что функции main,loop,setup действительно находятся во flash (с 0x08000000), и выполняются оттуда. Там gdb установить soft-breakpoint не может. В отличии от ARM где исполнимый код при старте копируется в специальную область ОЗУ. Поэтому мужик с ролика по отладке STM лихо расставлял точки останова.
Заметил такую странность. В пакете 80x-duino в файле csi_gcc.h присутствует такая конструкция:
`/**
  \brief   Breakpoint
  \details Causes the processor to enter Debug state
           Debug tools can use this to investigate system state when the instruction at a particular address is reached.
 */

__ALWAYS_STATIC_INLINE void __BKPT(void)
{
    __ASM volatile("bkpt");
}
` 
При наличии у процессора аппаратного триггера, он должен тормознуть на месте где написан __BKPT(); - и передать управление отладчику. Это не срабатывает. В этом заключается странность.

А так... все мы сделали правильно. c-sky gdb работает, прошивку можно трассировать. Но есть неудобство. Пока дошагаешь до Loop() - офигеешь :)

**Reply by AnatolSher** [reply]
Заработало! Оказывается можно через c-sky gdb устанавливать hardware breakpoint
Пошел раскуривать туториал на GDB :) 

![image](https://user-images.githubusercontent.com/121094544/218311030-fa8a3ec7-7b63-4c46-ae30-bb32e5fe86c3.png)

На этом этапе можно ставить точку. 
Второй этап - вкручивание этого хозяйства в Arduino IDE 2.x

**Comment by board707** [comment]
Поздравляю.
Будете разбираться с GDB - сразу пишите хелп файл. Сначала для меня, а потом мы его в репозиторий положим :)

**Comment by board707** [comment]
Внес обновления, касающиеся сборки S файлов, на гитхаб.

**Comment by AnatolSher** [comment]
Выяснил что пока в Arduino IDE 2x  встроить отладку других архитектур отличных от ARM пока не возможно. Это печально...но это пока. Зато для настоящих мужиков работает консольная отладка в Linux и весьма не плохо... Привыкнуть надо.

А для ботанов я запилил пакет для CDK под windows. Можно отлаживать как скетч так и ядро wiring.

![image](https://user-images.githubusercontent.com/121094544/218742230-1a35ab0b-08f0-47f2-9a3e-ee3c359db444.png)

Я пока не знаю какие возможности предоставляет CDK для управлениями вариантами. Но думаю дефайнами и препроцессорными директивами пока разруливать. Таким образом можно и под w801 настрогать расширений. имхо

**Reply by songzhishuo** [reply]
Can you give me this package, I think it will be more convenient to use CDK directly if you develop functions later

**Reply by AnatolSher** [reply]
See email

**Comment by eassyly** [comment]
can't INSTALL w80x_arduino csky plain
Note:

Downloading packages
w80x_duino:csky@2021.04.23
w80x_duino:w80x_tool@v1.0
w80x_duino:XT804@0.0.3
Installing w80x_duino:csky@2021.04.23
Failed to install platform: w80x_duino:XT804.
Error: 13 INTERNAL: Cannot install tool w80x_duino:csky@2021.04.23: searching package root dir: no unique root dir in archive, found 'C:\Users\eassy\AppData\Local\Arduino15\tmp\package-2833822628\bin' and 'C:\Users\eassy\AppData\Local\Arduino15\tmp\package-2833822628\csky-elfabiv2'

**Reply by board707** [reply]
Hi @eassyly !
Thanks for your inform.
Please try to install with a new json file https://github.com/board707/w80x_arduino/blob/hal-v0.6.0/package_w80x_test_index.json and report the results here or in issue #18
